util/iostat/iostat.go

```diff
@@ -0,0 +1,132 @@
+package iostat
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+)
+
+type MetricsSpawner struct {
+	statReceiver chan DeviceStats
+	metrics      map[string]map[string]metrics.GaugeFloat64
+}
+
+func NewMetricsSpawner() *MetricsSpawner {
+	return &MetricsSpawner{
+		metrics:      make(map[string]map[string]metrics.GaugeFloat64),
+		statReceiver: make(chan DeviceStats),
+	}
+}
+
+func (m *MetricsSpawner) RegisterMetrics(ctx context.Context, spwanInterval int) error {
+	go Run(ctx, spwanInterval, m.statReceiver)
+	// Register metrics for a maximum of 5 devices (fail safe incase iostat command returns incorrect names indefinitely)
+	for i := 0; i < 5; i++ {
+		stat, ok := <-m.statReceiver
+		if !ok {
+			return errors.New("failed to register iostat metrics")
+		}
+		if _, ok := m.metrics[stat.DeviceName]; ok {
+			return nil
+		}
+		baseMetricName := fmt.Sprintf("isotat/%s/", stat.DeviceName)
+		m.metrics[stat.DeviceName] = make(map[string]metrics.GaugeFloat64)
+		m.metrics[stat.DeviceName]["readspersecond"] = metrics.NewRegisteredGaugeFloat64(baseMetricName+"readspersecond", nil)
+		m.metrics[stat.DeviceName]["writespersecond"] = metrics.NewRegisteredGaugeFloat64(baseMetricName+"writespersecond", nil)
+		m.metrics[stat.DeviceName]["await"] = metrics.NewRegisteredGaugeFloat64(baseMetricName+"await", nil)
+	}
+	return nil
+}
+
+func (m *MetricsSpawner) PopulateMetrics() {
+	for {
+		stat, ok := <-m.statReceiver
+		if !ok {
+			log.Info("Iostat statReceiver channel was closed due to error or command being completed")
+			return
+		}
+		if _, ok := m.metrics[stat.DeviceName]; !ok {
+			log.Warn("Unrecognized device name in output of iostat command", "deviceName", stat.DeviceName)
+			continue
+		}
+		m.metrics[stat.DeviceName]["readspersecond"].Update(stat.ReadsPerSecond)
+		m.metrics[stat.DeviceName]["writespersecond"].Update(stat.WritesPerSecond)
+		m.metrics[stat.DeviceName]["await"].Update(stat.Await)
+	}
+}
+
+type DeviceStats struct {
+	DeviceName      string
+	ReadsPerSecond  float64
+	WritesPerSecond float64
+	Await           float64
+}
+
+func Run(ctx context.Context, interval int, receiver chan DeviceStats) {
+	defer close(receiver)
+	// #nosec G204
+	cmd := exec.Command("iostat", "-dNxy", strconv.Itoa(interval))
```

Comment:
I tried to run this command on MacOS, but it failed.
```
➜  ~ iostat -dNxy 1
iostat: illegal option -- N
usage: iostat [-CUdIKoT?] [-c count] [-n devs]
	      [-w wait] [drives]
```

I think this should be handled more gracefully either by explicitly disabling this feature or fixing the command for Mac.

Reply:
when the -dNxy flags aren't supported, the iostat command terminates and is being logged as an error.
the major use-case for these metrics is when the node is run within a container

Reply:
I'm proposing a simple check to replace the error message with a warning. Something like:
```
if goos.GOOS != "linux" {
    log.Warning("iostat unsuported")
} else {
    // try to initialize iostat
}
```

[goos](https://pkg.go.dev/internal/goos).[GOOS](https://pkg.go.dev/internal/goos#GOOS)